src/protocols/shell.rs

```diff
@@ -0,0 +1,292 @@
+//! Shell Protocol
+//!
+//! Provides shell services to UEFI applications.
+
+pub const PROTOCOL_GUID: crate::base::Guid = crate::base::Guid::from_fields(
+    0x6302d008,
+    0x7f9b,
+    0x4f30,
+    0x87,
+    0xac,
+    &[0x60, 0xc9, 0xfe, 0xf5, 0xda, 0x4e],
+);
+
+pub type ShellFileHandle = *mut core::ffi::c_void;
+
+pub type DeviceNameFlags = u32;
+pub const USE_COMPONENT_NAME: DeviceNameFlags = 0x00000001;
+pub const USE_DEVICE_PATH: DeviceNameFlags = 0x00000002;
+
+#[repr(C)]
+pub struct ListEntry {
+    pub flink: *mut ListEntry,
+    pub blink: *mut ListEntry,
+}
+
+#[repr(C)]
+pub struct ShellFileInfo {
+    pub link: ListEntry,
+    pub status: crate::base::Status,
+    pub full_name: *mut crate::base::Char16,
+    pub file_name: *mut crate::base::Char16,
+    pub handle: ShellFileHandle,
+    pub info: *mut crate::protocols::file::Info,
+}
+
+pub type Execute = eficall! {fn(
+    *mut crate::base::Handle,
+    *mut crate::base::Char16,
+    *mut *mut crate::base::Char16,
+    *mut crate::base::Status,
+) -> crate::base::Status};
+
+pub type GetEnv = eficall! {fn(
+    *mut crate::base::Char16,
+) -> *mut crate::base::Char16};
+
+pub type SetEnv = eficall! {fn(
+    *mut crate::base::Char16,
+    *mut crate::base::Char16,
+    crate::base::Boolean,
+) -> crate::base::Status};
+
+pub type GetAlias = eficall! {fn(
+    *mut crate::base::Char16,
+    *mut crate::base::Boolean,
+) -> *mut crate::base::Char16};
+
+pub type SetAlias = eficall! {fn(
+    *mut crate::base::Char16,
+    *mut crate::base::Char16,
+    crate::base::Boolean,
+    crate::base::Boolean,
+) -> crate::base::Status};
+
+pub type GetHelpText = eficall! {fn(
+    *mut crate::base::Char16,
+    *mut crate::base::Char16,
+    *mut *mut crate::base::Char16,
+) -> crate::base::Status};
+
+pub type GetDevicePathFromMap = eficall! {fn(
+    *mut crate::base::Char16,
+) -> *mut crate::protocols::device_path::Protocol};
+
+pub type GetMapFromDevicePath = eficall! {fn(
+    *mut *mut crate::protocols::device_path::Protocol,
+) -> *mut crate::base::Char16};
+
+pub type GetDevicePathFromFilePath = eficall! {fn(
+    *mut crate::base::Char16,
+) -> *mut crate::protocols::device_path::Protocol};
+
+pub type GetFilePathFromDevicePath = eficall! {fn(
+    *mut *mut crate::protocols::device_path::Protocol,
```

Comment:
This should be `*mut` rather than `*mut *mut`, right?

Reply:
> This should be `*mut` rather than `*mut *mut`, right?

Yes. Should I fix that myself?

Reply:
No, I already have it queued locally.